toolchain/check/eval.cpp

```diff
@@ -1687,6 +1687,64 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
       return context.constant_values().Get(arg_ids[0]);
     }
 
+    case SemIR::BuiltinFunctionKind::StringAt: {
+      Phase phase = Phase::Concrete;
+      auto str_id = GetConstantValue(eval_context, arg_ids[0], &phase);
+      auto index_id = GetConstantValue(eval_context, arg_ids[1], &phase);
+
+      if (phase != Phase::Concrete) {
+        return MakeNonConstantResult(phase);
+      }
+
+      auto str_struct = eval_context.insts().GetAs<SemIR::StructValue>(str_id);
+      auto elements = eval_context.inst_blocks().Get(str_struct.elements_id);
+      CARBON_CHECK(elements.size() == 2, "String struct should have 2 fields.");
+
+      auto ptr_const_id = eval_context.constant_values().Get(elements[0]);
+      auto string_literal = eval_context.insts().GetAs<SemIR::StringLiteral>(
+          eval_context.constant_values().GetInstId(ptr_const_id));
+
+      const auto& string_value =
+          eval_context.sem_ir().string_literal_values().Get(
+              string_literal.string_literal_id);
+
+      auto index_inst = eval_context.insts().GetAs<SemIR::IntValue>(index_id);
+      const auto& index_val = eval_context.ints().Get(index_inst.int_id);
+
+      if (index_val.isNegative()) {
+        CARBON_DIAGNOSTIC(StringIndexNegative, Error,
+                          "index `{0}` is negative.", TypedInt);
+        context.emitter().Emit(
+            loc_id, StringIndexNegative,
+            {.type = eval_context.insts().Get(arg_ids[1]).type_id(),
+             .value = index_val});
+        return SemIR::ConstantId::NotConstant;
+      }
+
+      if (index_val.getActiveBits() > 64 ||
+          index_val.getZExtValue() >= string_value.size()) {
+        CARBON_DIAGNOSTIC(
+            StringAtIndexOutOfBounds, Error,
+            "string index `{0}` is out of bounds; string has length {1}.",
+            TypedInt, unsigned);
+        context.emitter().Emit(
+            loc_id, StringAtIndexOutOfBounds,
+            {.type = eval_context.insts().Get(arg_ids[1]).type_id(),
```

Comment:
```suggestion
            {.type = eval_context.insts().Get(index_id).type_id(),
```
Should give us the same thing, right?